integration_test/exqlite/ecto_sql/migrator.exs

```diff
@@ -0,0 +1,243 @@
+    assert migrated_versions(PoolRepo) == []
+  end
+
+  @tag :prefix
+  test "does not commit migration if insert into schema migration fails" do
+    # First we create a new schema migration table in another prefix
+    assert up(PoolRepo, 33, AnotherSchemaMigration, log: false) == :ok
+    assert migrated_versions(PoolRepo) == [33]
+
+    catch_error(up(PoolRepo, 34, GoodMigration, log: false, prefix: "bad_schema_migrations"))
+    catch_error(PoolRepo.all("good_migration"))
+    catch_error(PoolRepo.all("good_migration", prefix: "bad_schema_migrations"))
+
+    assert down(PoolRepo, 33, AnotherSchemaMigration, log: false) == :ok
+  end
+
+  test "bad execute migration" do
+    assert catch_error(up(PoolRepo, 31, BadMigration, log: false))
+  end
+
+  test "broken link migration" do
+    Process.flag(:trap_exit, true)
+
+    assert capture_log(fn ->
+      {:ok, pid} = Task.start_link(fn -> up(PoolRepo, 31, BrokenLinkMigration, log: false) end)
+      assert_receive {:EXIT, ^pid, _}
+    end) =~ "oops"
+
+    assert capture_log(fn ->
+      catch_exit(up(PoolRepo, 31, BrokenLinkMigration, log: false))
+    end) =~ "oops"
+  end
+
+  test "run up to/step migration", config do
+    in_tmp fn path ->
+      create_migration(47, config)
+      create_migration(48, config)
+
+      assert [47] = run(PoolRepo, path, :up, step: 1, log: false)
+      assert count_entries() == 1
+
+      assert [48] = run(PoolRepo, path, :up, to: 48, log: false)
+    end
+  end
+
+  test "run down to/step migration", config do
+    in_tmp fn path ->
+      migrations = [
+        create_migration(49, config),
+        create_migration(50, config),
+      ]
+
+      assert [49, 50] = run(PoolRepo, path, :up, all: true, log: false)
+      purge migrations
+
+      assert [50] = run(PoolRepo, path, :down, step: 1, log: false)
+      purge migrations
+
+      assert count_entries() == 1
+      assert [50] = run(PoolRepo, path, :up, to: 50, log: false)
+    end
+  end
+
+  test "runs all migrations", config do
+    in_tmp fn path ->
+      migrations = [
+        create_migration(53, config),
+        create_migration(54, config),
+      ]
+
+      assert [53, 54] = run(PoolRepo, path, :up, all: true, log: false)
+      assert [] = run(PoolRepo, path, :up, all: true, log: false)
+      purge migrations
+
+      assert [54, 53] = run(PoolRepo, path, :down, all: true, log: false)
+      purge migrations
+
+      assert count_entries() == 0
+      assert [53, 54] = run(PoolRepo, path, :up, all: true, log: false)
+    end
+  end
+
+  test "does not commit half transactions on bad syntax", config do
+    in_tmp fn path ->
+      migrations = [
+        create_migration(64, config),
+        create_migration("65_+", config)
+      ]
+
+      assert_raise SyntaxError, fn ->
+        run(PoolRepo, path, :up, all: true, log: false)
+      end
+
+      refute_received {:up, _}
+      assert count_entries() == 0
+      purge migrations
+    end
+  end
+
+  @tag :lock_for_migrations
```

Comment:
added exclusion here

Reply:
Should probably push this upstream as well